pkg/storage/redis.go

```diff
@@ -0,0 +1,366 @@
+	}
+
+	if err := b.mutex.Lock(); err != nil {
+		return errors.Wrap(err, "cannot obtain mutex lock")
+	}
+	defer func() {
+		ok, unlockErr := b.mutex.Unlock()
+		if !ok || unlockErr != nil {
+			logrus.Error(unlockErr)
+		}
+	}()
+
+	// The Pipeliner interface provided by the go-redis library guarantees that all the commands queued in the pipeline will either succeed or fail together.
+	_, err := b.db.TxPipelined(b.ctx, func(pipe goredislib.Pipeliner) error {
+		for i := range namespaces {
+			namespace := namespaces[i]
+			key := keys[i]
+			value := values[i]
+
+			nameSpaceKey := getRedisKey(namespace, key)
+			err := pipe.Set(b.ctx, nameSpaceKey, value, 0).Err()
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+func (b *RedisDB) Read(namespace, key string) ([]byte, error) {
+	nameSpaceKey := getRedisKey(namespace, key)
+	res, err := b.db.Get(b.ctx, nameSpaceKey).Bytes()
+
+	// Nil reply returned by Redis when key does not exist.
+	if errors.Is(err, goredislib.Nil) {
+		return res, nil
+	}
+
+	return res, err
+}
+
+func (b *RedisDB) ReadPrefix(namespace, prefix string) (map[string][]byte, error) {
+	namespacePrefix := getRedisKey(namespace, prefix)
+
+	keys, err := readAllKeys(namespacePrefix, b)
+	if err != nil {
+		return nil, errors.Wrap(err, "read all keys error")
+	}
+
+	return readAll(keys, b)
+}
+
+func (b *RedisDB) ReadAll(namespace string) (map[string][]byte, error) {
+	keys, err := readAllKeys(namespace, b)
+	if err != nil {
+		return nil, errors.Wrap(err, "read all keys error")
+	}
+
+	return readAll(keys, b)
+}
+
+// TODO: This potentially could dangerous as it might run out of memory as we populate result
+func readAll(keys []string, b *RedisDB) (map[string][]byte, error) {
+	result := make(map[string][]byte)
+
+	if len(keys) == 0 {
+		return nil, nil
+	}
+
+	values, err := b.db.MGet(b.ctx, keys...).Result()
+	if err != nil {
+		return nil, errors.Wrap(err, "getting multiple keys")
+	}
+
+	if len(keys) != len(values) {
+		return nil, errors.New("key length does not match value length")
+	}
+
+	// result needs to take the namespace out of the key
+	namespaceDashIndex := strings.Index(keys[0], "-")
+	for i, val := range values {
+		byteValue := []byte(fmt.Sprintf("%v", val))
+		key := keys[i]
+		key = key[namespaceDashIndex+1:]
+		result[key] = byteValue
+	}
+
+	return result, nil
+}
+
+func (b *RedisDB) ReadAllKeys(namespace string) ([]string, error) {
+	keys, err := readAllKeys(namespace, b)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(keys) == 0 {
+		return make([]string, 0), nil
+	}
+
+	namespaceDashIndex := strings.Index(keys[0], "-")
+	for i, key := range keys {
+		keyWithoutNamespace := key[namespaceDashIndex+1:]
+		keys[i] = keyWithoutNamespace
+	}
+
+	return keys, nil
+}
+
+// TODO: This potentially could dangerous as it might run out of memory as we populate allKeys
+func readAllKeys(namespace string, b *RedisDB) ([]string, error) {
```

Comment:
Seems like you're removing the namespace out of the keys every time. Would it make sense to do it inside this function?